### PR TITLE
Workaround audio sample rate stream issue

### PR DESCRIPTION
**Pull request by SylvainMorel**
We are having sporadic issues with streams having inconsistent audio sample rates. The container and the data are at 48 kHz, but the atom dac3 has the sample rate at 32 kHz. That messes up with the sync mechanisms and breaks the playback.

This PR adds a workaround the issue, forcing the dac3 output to 48 kHz when it's inconsistent with the container (can be enabled by setting a global flag in Util, disabled by default)

Also added a new function in Log, to be able to log an exception, which allows us to notify the app when the sample rate error happens.

Also kept some of the logs improvements I made during the investigation.

### CODE REVIEW

File: library/core/src/main/java/com/google/android/exoplayer2/audio/MediaCodecAudioRenderer.java [hunk 1]
```diff
@@ -689,8 +689,8 @@ protected boolean processOutputBuffer(
       throws ExoPlaybackException {
     checkNotNull(buffer);
 
-    Log.v(Log.LOG_LEVEL_VERBOSE2, TAG, "processOutputBuffer (decryptOnlyCodecFormat %s  bufferFlags: %d isDecodeOnlyBuffer %s)",
-        decryptOnlyCodecFormat, bufferFlags, isDecodeOnlyBuffer);
+    Log.v(Log.LOG_LEVEL_VERBOSE2, TAG, "processOutputBuffer prezTime: %d sampleCount: %d (decryptOnlyCodecFormat %s  bufferFlags: %d isDecodeOnlyBuffer %s)",
```
Comment (flambert) [inline library/core/src/main/java/com/google/android/exoplayer2/audio/MediaCodecAudioRenderer.java:692]:
```suggestion
    Log.v(Log.LOG_LEVEL_VERBOSE2, TAG, "processOutputBuffer presentationTime: %d sampleCount: %d (decryptOnlyCodecFormat %s  bufferFlags: %d isDecodeOnlyBuffer %s)",
```

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java [hunk 2]
```diff
@@ -1536,6 +1537,12 @@ private static void parseAudioSampleEntry(
         parent.setPosition(Atom.HEADER_SIZE + childPosition);
         out.format =
             Ac3Util.parseAc3AnnexFFormat(parent, Integer.toString(trackId), language, drmInitData);
+
+        // MIREGO added workaround for stream data issue where sample rate is inconsistent
+        if (Util.shouldWorkaroundAudioSampleRateDataBug && (out.format.sampleRate == 32000) && (sampleRate == 48000)) {
+          out.format = out.format.buildUpon().setSampleRate(sampleRate).build();
+          Log.e(TAG, new PlaybackException("Inconsistent audio sample rate", new RuntimeException(), PlaybackException.ERROR_CODE_AUDIO_TRACK_INCONSISTENT_SAMPLE_RATE));
```
Comment (flambert) [inline library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java:1544]: Ca donne de quoi de plus de passer une RuntimeException ou bien c'est parce que t'as pas le choix, ca a besoin d'une "reason" exception?

**Review by flambert — APPROVED** [review]

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java (hunk 2 above)
Comment (SylvainMorel) [reply library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java:1544]: Ouin, pas le choix, le constructeur est fait de même...

**Review by jdtremblay — COMMENTED** [review]
late 👍

**Pull request merged**